File: desktop/macos-saver/PearWallRuntimeState.c
#include <stdatomic.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>

enum {
    PEARWALL_RUNTIME_STATE_VERSION = 1,
    PEARWALL_RUNTIME_STATE_SIZE = 64,
};

typedef struct {
    uint8_t magic[8];
    uint32_t version;
    uint32_t size;
    _Atomic uint32_t sequence;
    _Atomic uint32_t pulse_bits;
    _Atomic uint32_t playing;
    uint32_t reserved_value;
    _Atomic uint64_t updated_at_milliseconds;
    _Atomic uint64_t settings_revision;
    uint8_t reserved[16];
} PearWallRuntimeState;

_Static_assert(sizeof(PearWallRuntimeState) == PEARWALL_RUNTIME_STATE_SIZE, "invalid runtime state size");

typedef struct {
    const PearWallRuntimeState *state;
} PearWallRuntimeStateReader;

static const uint8_t pearwall_runtime_magic[8] = {
    'P', 'W', 'R', 'S', 'T', 'A', 'T', 'E',
};

void *pearwall_runtime_state_open(const char *path) {
    if (path == NULL) {
        return NULL;
    }
    int descriptor = open(path, O_RDONLY | O_CLOEXEC);
    if (descriptor < 0) {
        return NULL;
    }
    struct stat attributes;
    if (fstat(descriptor, &attributes) != 0 ||
        attributes.st_size < PEARWALL_RUNTIME_STATE_SIZE) {
        close(descriptor);
        return NULL;
    }
    void *mapping = mmap(
        NULL,
        PEARWALL_RUNTIME_STATE_SIZE,
        PROT_READ,
        MAP_SHARED,
        descriptor,
        0
    );
    close(descriptor);
    if (mapping == MAP_FAILED) {
        return NULL;
    }
    const PearWallRuntimeState *state = mapping;
    if (memcmp(state->magic, pearwall_runtime_magic, sizeof(pearwall_runtime_magic)) != 0 ||
        state->version != PEARWALL_RUNTIME_STATE_VERSION ||
        state->size != PEARWALL_RUNTIME_STATE_SIZE) {
        munmap(mapping, PEARWALL_RUNTIME_STATE_SIZE);
        return NULL;
    }
    PearWallRuntimeStateReader *reader = calloc(1, sizeof(*reader));
    if (reader == NULL) {
        munmap(mapping, PEARWALL_RUNTIME_STATE_SIZE);
        return NULL;
    }
    reader->state = state;
    return reader;
}

int32_t pearwall_runtime_state_read(
    void *opaque_reader,
    float *pulse,
    uint64_t *updated_at_milliseconds
) {
    if (opaque_reader == NULL || pulse == NULL || updated_at_milliseconds == NULL) {
        return 0;
    }
    const PearWallRuntimeStateReader *reader = opaque_reader;
    const PearWallRuntimeState *state = reader->state;
    for (int attempt = 0; attempt < 8; attempt += 1) {
        uint32_t first = atomic_load_explicit(&state->sequence, memory_order_acquire);
        if ((first & 1U) != 0) {
            continue;
        }
        uint32_t pulse_bits = atomic_load_explicit(&state->pulse_bits, memory_order_relaxed);
        uint64_t updated = atomic_load_explicit(
            &state->updated_at_milliseconds,
            memory_order_relaxed
        );
        uint32_t second = atomic_load_explicit(&state->sequence, memory_order_acquire);
        if (first != second || (second & 1U) != 0) {
            continue;
        }
        memcpy(pulse, &pulse_bits, sizeof(pulse_bits));
        *updated_at_milliseconds = updated;
        return 1;
    }
    return 0;
}

void pearwall_runtime_state_close(void *opaque_reader) {
    if (opaque_reader == NULL) {
        return;
    }
    PearWallRuntimeStateReader *reader = opaque_reader;
    munmap((void *)reader->state, PEARWALL_RUNTIME_STATE_SIZE);
    free(reader);
}

File: desktop/plugins/pearwall-wallpaper/macos/Sources/PearWallWallpaperNative/Core/PearWallRuntimeState.swift
import Darwin
import Foundation

struct PearWallRuntimeSnapshot {
    let pulse: Float
    let playing: Bool
    let playbackRate: Float
    let trackID: UInt64
    let position: Double
    let duration: Double
    let updatedAtMilliseconds: UInt64
}

final class PearWallRuntimeStateReader {
    private static let stateSize = 128
    private static let maximumAgeMilliseconds: UInt64 = 1_000
    private static let magic: [UInt8] = Array("PWRSTATE".utf8)
    private var mapping: UnsafeMutableRawPointer?
    private var nextOpenAttempt = Date.distantPast
    private let stateURL: URL?

    init(stateURL: URL? = nil) {
        self.stateURL = stateURL
    }

    deinit {
        closeMapping()
    }

    func currentSnapshot() -> PearWallRuntimeSnapshot? {
        openIfNeeded()
        guard let mapping else { return nil }
        for _ in 0..<8 {
            let first = loadUInt32(mapping, offset: 16)
            if first & 1 != 0 {
                continue
            }
            OSMemoryBarrier()
            let pulseBits = loadUInt32(mapping, offset: 20)
            let playing = loadUInt32(mapping, offset: 24) != 0
            let playbackRateBits = loadUInt32(mapping, offset: 28)
            let updatedAtMilliseconds = loadUInt64(mapping, offset: 32)
            let trackID = loadUInt64(mapping, offset: 48)
            let positionBits = loadUInt64(mapping, offset: 56)
            let durationBits = loadUInt64(mapping, offset: 64)
            OSMemoryBarrier()
            let second = loadUInt32(mapping, offset: 16)
            let pulse = Float(bitPattern: pulseBits)
            let playbackRate = Float(bitPattern: playbackRateBits)
            let position = Double(bitPattern: positionBits)
            let duration = Double(bitPattern: durationBits)
            guard first == second,
                  second & 1 == 0,
                  pulse.isFinite,
                  playbackRate.isFinite,
                  position.isFinite,
                  duration.isFinite else {
                continue
            }
            let now = UInt64(max(0, Date().timeIntervalSince1970 * 1_000))
            guard now >= updatedAtMilliseconds,
                  now - updatedAtMilliseconds <= Self.maximumAgeMilliseconds else {
                closeMapping(retryAfter: 1)
                return nil
            }
            return PearWallRuntimeSnapshot(
                pulse: min(1, max(0, pulse)),
                playing: playing,
                playbackRate: max(0, playbackRate),
                trackID: trackID,
                position: max(0, position),
                duration: max(0, duration),
                updatedAtMilliseconds: updatedAtMilliseconds,
            )
        }
        return nil
    }

    private func openIfNeeded() {
        guard mapping == nil, Date() >= nextOpenAttempt else { return }
        nextOpenAttempt = Date().addingTimeInterval(1)
        guard let url = stateURL ?? pearWallApplicationSupportDirectory()?
            .appendingPathComponent("runtime-state-v2.bin") else {
            return
        }
        let descriptor = Darwin.open(url.path, O_RDONLY | O_CLOEXEC)
        guard descriptor >= 0 else { return }
        defer { Darwin.close(descriptor) }
        var attributes = stat()
        guard fstat(descriptor, &attributes) == 0,
              attributes.st_size >= Self.stateSize else {
            return
        }
        let candidate = mmap(nil, Self.stateSize, PROT_READ, MAP_SHARED, descriptor, 0)
        guard candidate != MAP_FAILED,
              let candidate,
              validState(candidate) else {
            if candidate != MAP_FAILED, let candidate {
                munmap(candidate, Self.stateSize)
            }
            return
        }
        mapping = candidate
    }

    private func closeMapping(retryAfter: TimeInterval = 0) {
        guard let mapping else { return }
        munmap(mapping, Self.stateSize)
        self.mapping = nil
        nextOpenAttempt = Date().addingTimeInterval(retryAfter)
    }

    private func validState(_ mapping: UnsafeMutableRawPointer) -> Bool {
        let validMagic = Self.magic.withUnsafeBytes { bytes in
            memcmp(mapping, bytes.baseAddress, Self.magic.count) == 0
        }
        return validMagic
            && loadUInt32(mapping, offset: 8) == 2
            && loadUInt32(mapping, offset: 12) == Self.stateSize
    }

    private func loadUInt32(_ pointer: UnsafeMutableRawPointer, offset: Int) -> UInt32 {
        UInt32(littleEndian: pointer.load(fromByteOffset: offset, as: UInt32.self))
    }

    private func loadUInt64(_ pointer: UnsafeMutableRawPointer, offset: Int) -> UInt64 {
        UInt64(littleEndian: pointer.load(fromByteOffset: offset, as: UInt64.self))
    }
}
